### PR TITLE
Adding support for elementary OS distro in install script.

### PR DESCRIPTION
**Pull request by fmoliveira**
Adding support for elementary OS in the installer script.

The script will query the upstream release from the command `$ lsb_release -a -u`, so, it's now very easy to add support for more forked distros.

~~I don't know what are the side effects of checking the upstream release for all distros. For that reason, I've restricted this check to elementary OS, which I've been able to test.~~

I've tested this in elementary OS 0.3 (freya) and it's working just fine.

Credits to Daniel Foré and Cody Garver from elementary OS for [helping me out to achieve this](http://elementaryos.stackexchange.com/a/769).

**TL;DR:**
We've tested the command in some distros (based and non-based in Debian/Ubuntu) and have achieved a solution that should work across all distros. It'll check for the existince and success of the command `lsb_release -a -u` to then retrieve the upstream release info.

### CODE REVIEW

**Comment by fmoliveira** [comment]
What do you think of this @jfrazelle ?

**Comment by jfrazelle** [comment]
sweet this is neat!, ping @tianon

**Comment by jfrazelle** [comment]
LGTM, thanks for finding this out, its seriously going to make things so much cleaner

**Comment by fmoliveira** [comment]
No problem! :smiley: 

I've edited the first message to give credits to the folks from elementary OS for helping me out. :+1:

File: hack/install.sh [hunk 1]
```diff
@@ -51,6 +51,23 @@ echo_docker_as_nonroot() {
 	EOF
 }
 
+# Check upstream release in forked Linux distros
+check_forked() {
+	if [ "$lsb_dist" = "$1" ]; then
```
Comment (benmccann) [inline hack/install.sh]: We'll also want to check for linuxmint, cumulus linux, kali, etc. here. Right now, if we want to check multiple distros we have to call check_forked multiple times. I'm not sure if that's the cleanest way? Other thoughts I have would be to either hardcode the list in this if check or perhaps even better if it works across distros would be to remove the if check entirely and always find lsd_dist and dist_version the way that's being done here

**Comment by fmoliveira** [comment]
@jfrazelle @benmccann 

Surely it'd be better if it works across all distros. Let's test in all platforms we can.

**Comment by fmoliveira** [comment]
## How to test your platform
Run the following commands and post their respective outputs here:
```
$ lsb_release -a
$ lsb_release -a -u
```

**Comment by fmoliveira** [comment]
For elementary OS 0.3:
```
fmoliveira:~ $ lsb_release -a
No LSB modules are available.
Distributor ID:	elementary OS
Description:	elementary OS Freya
Release:	0.3
Codename:	freya

fmoliveira:~ $ lsb_release -a -u
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04 LTS
Release:	14.04
Codename:	trusty
```

**Comment by fmoliveira** [comment]
Sorry I've accidentaly closed the pull request. Already reopened.

**Comment by fmoliveira** [comment]
For Ubuntu 14.04:
```
filipe:~ $ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.2 LTS
Release:	14.04
Codename:	trusty

filipe:~ $ lsb_release -a -u
Usage: lsb_release [options]

lsb_release: error: no such option: -u

filipe:~ $ echo $?
2

filipe:~ $ ls /etc/upstream-release/
ls: cannot access /etc/upstream-release/: No such file or directory

filipe:~ $ cat /etc/upstream-release/*
cat: /etc/upstream-release/*: No such file or directory
```

Please note as the `-u` argument is unsupported in Ubuntu, the command will put an error number to exit code. This might be a safe way to determine whether the user is running a forked distro or not.

**Comment by benmccann** [comment]
For Linux Mint:

    $ lsb_release -a
    LSB Version:	core-2.0-amd64:core-2.0-noarch:core-3.0-amd64:core-3.0-noarch:core-3.1-amd64:core-3.1-noarch:core-3.2-amd64:core-3.2-noarch:core-4.0-amd64:core-4.0-noarch:core-4.1-amd64:core-4.1-noarch:cxx-3.0-amd64:cxx-3.0-noarch:cxx-3.1-amd64:cxx-3.1-noarch:cxx-3.2-amd64:cxx-3.2-noarch:cxx-4.0-amd64:cxx-4.0-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-3.1-amd64:desktop-3.1-noarch:desktop-3.2-amd64:desktop-3.2-noarch:desktop-4.0-amd64:desktop-4.0-noarch:desktop-4.1-amd64:desktop-4.1-noarch:graphics-2.0-amd64:graphics-2.0-noarch:graphics-3.0-amd64:graphics-3.0-noarch:graphics-3.1-amd64:graphics-3.1-noarch:graphics-3.2-amd64:graphics-3.2-noarch:graphics-4.0-amd64:graphics-4.0-noarch:graphics-4.1-amd64:graphics-4.1-noarch:languages-3.2-amd64:languages-3.2-noarch:languages-4.0-amd64:languages-4.0-noarch:languages-4.1-amd64:languages-4.1-noarch:multimedia-3.2-amd64:multimedia-3.2-noarch:multimedia-4.0-amd64:multimedia-4.0-noarch:multimedia-4.1-amd64:multimedia-4.1-noarch:printing-3.2-amd64:printing-3.2-noarch:printing-4.0-amd64:printing-4.0-noarch:printing-4.1-amd64:printing-4.1-noarch:qt4-3.1-amd64:qt4-3.1-noarch:security-4.0-amd64:security-4.0-noarch:security-4.1-amd64:security-4.1-noarch
    Distributor ID:	LinuxMint
    Description:	Linux Mint 17.2 Rafaela
    Release:	17.2
    Codename:	rafaela
    
    $ lsb_release -a -u
    LSB Version:	core-2.0-amd64:core-2.0-noarch:core-3.0-amd64:core-3.0-noarch:core-3.1-amd64:core-3.1-noarch:core-3.2-amd64:core-3.2-noarch:core-4.0-amd64:core-4.0-noarch:core-4.1-amd64:core-4.1-noarch:cxx-3.0-amd64:cxx-3.0-noarch:cxx-3.1-amd64:cxx-3.1-noarch:cxx-3.2-amd64:cxx-3.2-noarch:cxx-4.0-amd64:cxx-4.0-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-3.1-amd64:desktop-3.1-noarch:desktop-3.2-amd64:desktop-3.2-noarch:desktop-4.0-amd64:desktop-4.0-noarch:desktop-4.1-amd64:desktop-4.1-noarch:graphics-2.0-amd64:graphics-2.0-noarch:graphics-3.0-amd64:graphics-3.0-noarch:graphics-3.1-amd64:graphics-3.1-noarch:graphics-3.2-amd64:graphics-3.2-noarch:graphics-4.0-amd64:graphics-4.0-noarch:graphics-4.1-amd64:graphics-4.1-noarch:languages-3.2-amd64:languages-3.2-noarch:languages-4.0-amd64:languages-4.0-noarch:languages-4.1-amd64:languages-4.1-noarch:multimedia-3.2-amd64:multimedia-3.2-noarch:multimedia-4.0-amd64:multimedia-4.0-noarch:multimedia-4.1-amd64:multimedia-4.1-noarch:printing-3.2-amd64:printing-3.2-noarch:printing-4.0-amd64:printing-4.0-noarch:printing-4.1-amd64:printing-4.1-noarch:qt4-3.1-amd64:qt4-3.1-noarch:security-4.0-amd64:security-4.0-noarch:security-4.1-amd64:security-4.1-noarch
    Distributor ID:	Ubuntu
    Description:	Ubuntu 14.04 LTS
    Release:	14.04
    Codename:	trusty
    
    $ ls /etc/upstream-release/
    lsb-release
    
    $ cat /etc/upstream-release/*
    DISTRIB_ID=Ubuntu
    DISTRIB_RELEASE=14.04
    DISTRIB_CODENAME=trusty
    DISTRIB_DESCRIPTION="Ubuntu 14.04 LTS"

**Comment by fmoliveira** [comment]
Thanks @benmccann ! The upstream release info will make the script work perfectly in Linux Mint.

I'll create some virtual machines with other distros to test them.

**Comment by fmoliveira** [comment]
For Debian Jessie:
```
user@debian:~ $ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.1 (jessie)
Release:        8.1
Codename:       jessie

user@debian:~ $ lsb_release -a -u
Usage: lsb_release [options]

lsb_release: error: no such option: -u

user@debian:~ $ echo $?
2

user@debian:~ $ ls /etc/upstream-release/
ls: cannot access /etc/upstream-release/: No such file or directory

user@debian:~ $ cat /etc/upstream-release/*
cat: /etc/upstream-release/*: No such file or directory
```

**Comment by fmoliveira** [comment]
For Kubuntu 14.04:
```
kubuntu@kubuntu:~ $ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.2 LTS
Release:        14.04
Codename:       trusty

kubuntu@kubuntu:~ $ lsb_release -a -u
Usage: lsb_release [options]

lsb_release: error: no such option: -u

kubuntu@kubuntu:~ $ echo $?
2

kubuntu@kubuntu:~ $ ls /etc/upstream-release/
ls: cannot access /etc/upstream-release/: No such file or directory

kubuntu@kubuntu:~ $ cat /etc/upstream-release/*
cat: /etc/upstream-release/*: No such file or directory
```

**Comment by fmoliveira** [comment]
For openSUSE:
```
linux@linux:~> lsb_release -a
LSB Version:	n/a
Distributor ID:	openSUSE project
Description:	openSUSE 13.2 (Harlequin) (x86_64)
Release:	    13.2
Codename:	    Harlequin

linux@linux:~> lsb_release -a -u
lsb_release: invalid option -- 'u'

linux@linux:~> echo $?
2

linux@linux:~> ls /etc/upstream-release/
ls: cannot access /etc/upstream-release/: No such file or directory
linux@linux:~> cat /etc/upstream-release/*

cat: /etc/upstream-release/*: No such file or directory
```

**Comment by fmoliveira** [comment]
RedHat based distros don't usually have the `lsb_release` command installed by default, though it's possible to install it.

I haven't tested the command in any RH based distro, but I'm considering the output will be similar to openSUSE's, which isn't Debian based too.

**Comment by fmoliveira** [comment]
## Conclusion
Well, I think we've had enough tests for a good conclusion.

Some distros don't have the `lsb_release` installed by default. But everything based on Debian seems to have it. The base Debian, Ubuntu and Kubuntu distros don't have the `-u` argument to print the upstream release, and an error is raised in the exit code of the command. It seems we can trust the command in all distros when it returns zero (no error).

We can do this then:
 1. Check if command `lsb_release` exists.
 2. Try to set a variable with the output of the command `lsb_release -a -u`.
 3. If zero is returned, it's safe to assume we've got the Distribution ID and Codename from the upstream distro and we can replace that info in the script variables.
 4. If non-zero is returned, we'll do not touch the variables and let the script run as it would do it today.

I'll change the script and amend this pull request.

**Comment by fmoliveira** [comment]
I've amended the changes to the script, following the logic from the "conclusion" comment.

I've also improved the scripting sequence a little bit.

It's working fine in elementary OS and so should too in any other distros. Tests from people using other distros are always welcome anyway. :smiley:

**Comment by fmoliveira** [comment]
I've amended the changes to the script, following the logic from the "conclusion" comment.

I've also improved the scripting sequence a little bit.

It's working fine in elementary OS and so should too in any other distros. Tests from people using other distros are always welcome anyway. :smiley:

**Comment by benmccann** [comment]
very awesome. this looks good to me

**Comment by Djelibeybi** [comment]
:-1: This will break on Red Hat and derivatives. The `-u` parameter is invalid:

    $ lsb_release -a -u
    lsb_release: invalid option -- 'u'

**Comment by fmoliveira** [comment]
Someone posted a comment here with the following message, but it's already been deleted:
```
This will break on Red Hat and derivatives. The -u parameter is invalid:
```
Whoever it was, I'm aware that major distros don't have the `-u` option and the script is prepared to handle that scenario. Please note we were discussing this in the previous comments to avoid breaking the script in any distro.

Note the following line of code:
```
if [ "$?" = "0" ]; then
```
It'll check if the previous command exited with no error before setting new values to `lsb_dist` and `dist_version`.

Anyway, we'd be glad if you could test the modified script from my repo and tell us if an error really ocurred in your distribution. Please try either of the following commands:
```
wget -qO- https://raw.githubusercontent.com/fmoliveira/docker/master/hack/install.sh | sh
```
or:
```
curl -o- https://raw.githubusercontent.com/fmoliveira/docker/master/hack/install.sh | sh
```
If you've any further questions please do not hesitate posting here. We're to help. :+1:

**Comment by fmoliveira** [comment]
**One point for improvement:** We could do all the string manipulation stuff after this validation for better readability, instead of doing both the check and the string manipulation in one line. I'll do that and amend this commit.

**Comment by fmoliveira** [comment]
Someone posted a comment here with the following message, but it's already been deleted:
```
This will break on Red Hat and derivatives. The -u parameter is invalid:
```
Surely we're concerned about major distros. We're aware that major distros don't have the `-u` option and the script is prepared to handle that scenario.

Please note the following line of code:
```
if [ "$?" = "0" ]; then
```
It'll check if the previous command exited with no error before setting new values to `lsb_dist` and `dist_version`.

Anyway, we'd be glad if you could test the modified script from my repo and tell us if an error really ocurred in your distribution. Please try either of the following commands:
```
wget -qO- https://raw.githubusercontent.com/fmoliveira/docker/master/hack/install.sh | sh
```
or:
```
curl -o- https://raw.githubusercontent.com/fmoliveira/docker/master/hack/install.sh | sh
```
If you've any further questions please do not hesitate posting here. We're to help. :+1:

**Comment by fmoliveira** [comment]
**One point for improvement:** We could do all the string manipulation stuff after this validation for better readability, instead of doing both the check and the string manipulation in one line. I'll do that and amend this commit.

**Comment by fmoliveira** [comment]
Based in the last comment, I've seen a point for improvement.

Instead of doing the `-u` check and string manipulation all in one line, we should split them in separate lines to avoid piping the error to the next commands. This will automatically improve readability.

I'll do that and amend this pull request.

**Comment by fmoliveira** [comment]
I've amended this pull request with changes suggested in the previous comment.

The `lsb_release -a -u` command will be ran once to verify if the options exists, before proceeding to format the output.

Also, the script is not setting a variable with the command output anymore because it'd remove newlines and break grep usage. In the previous commit it relied in replacing newlines with separators and cutting them by indexes, but it'll break if the order of Distribution ID and Codename in the command output changes someday.

**Comment by fmoliveira** [comment]
I've amended this pull request with changes suggested in the previous comment.

The `lsb_release -a -u` command will be ran once to verify if the options exists, before proceeding to format the output. This will avoid piping the error to the next commands.

Also, the script is not setting a variable with the command output anymore because it'd remove newlines and break grep usage. In the previous commit it relied in replacing newlines with separators and cutting them by indexes, but it'll break if the order of Distribution ID and Codename in the command output changes someday.

**Comment by Djelibeybi** [comment]
@fmoliveira Yes, I commented, but deleted it as soon as I realised you'd already dealt with the issue. My bad for commenting without properly reviewing the code. :+1:

**Comment by fmoliveira** [comment]
Oh, I see @Djelibeybi ! No problem! Extra caution is always good, especially when dealing with an important script such as the installer itself!

Thanks for pointing your concerns anyway, you've made me think in the latter improvement that I've amended. :wink:  :+1:

**Comment by fabioxgn** [comment]
Worked great on my eOS Freya, thanks. I was getting the "not easily detectable" error message.

:shipit:

**Comment by arner** [comment]
Also working on my Linux Mint 17.2 Rafaela. Thanks!

**Comment by thaJeztah** [comment]
@fmoliveira looks like this needs a rebase

**Comment by fmoliveira** [comment]
Ok @thaJeztah I'll do that!

**Comment by sudohippie** [comment]
Seems like Linux Mint 16 Petra doesn't have the ```-u``` option. Does this mean, it won't be install-able?

```
$ lsb_release -a -u
Usage: lsb_release [options]

lsb_release: error: no such option: -u

```

**Comment by fmoliveira** [comment]
Thanks for warning @sudohippie , I'll create a VM to test what we can do for Petra.

**Comment by tiborvass** [comment]
@fmoliveira sorry, needs a rebase!

**Comment by fmoliveira** [comment]
Ok @tiborvass , I'll do that very soon and amend this pull request. Thanks!

**Comment by fmoliveira** [comment]
@sudohippie @thaJeztah @tiborvass 

I've created a Mint 16 VM and the `-u` option really doesn't exist. I'm trying to find an alternative solution for this case.

I'll analyse this before rebasing my pull request.

**Comment by benmccann** [comment]
@fmoliveira can you fall back to the solution in https://github.com/docker/docker/pull/14991 on Linux Mint 16? I would test, but only have Mint 17.2 installed currently.

**Comment by sudohippie** [comment]
@fmoliveira, @benmccann  - Looking at the download-able [versions](http://www.linuxmint.com/download_all.php) on Linux Mint website, it seems like Petra 16 has been discontinued. Not sure if this should be supported. I am planning to move to 17.2 now.

**Comment by thaJeztah** [comment]
> I'll analyse this issue before rebasing my pull request

Alright. Maybe should consider merging this PR at some point, and improve further in a follow-up.

> it seems like Petra 16 has been discontinued. Not sure if this should be supported

Good point; we should definitely be careful not to try and fix *every* version out there, especially if it's deprecated. Users can still add the repo manually.


ping @tianon for thoughts (perhaps ideas for better handling of the `-u`) :smile: (would love to see this move forward :+1:)

**Comment by fmoliveira** [comment]
@benmccann Unfortunately the file `/etc/upstream-release/lsb-release` or directory `/etc/upstream-release` as mentioned in #14991 doesn't exist in Mint 16.

**Comment by fmoliveira** [comment]
@thaJeztah I get your point. You're absolutely right, we'll already improve support for several distros.

In Mint 16 the script result just won't change, but surely we could improve further if needed. Good point at Mint 16 being discontinued.

I'll rebase this today and amend the PR.

**Comment by thaJeztah** [comment]
@fmoliveira :+1: great! Fwiw, @tianon is "leading" in this area, but I just wanted to avoid having this PR open for weeks, only to 'tweak' things (while it is already a good improvement)

**Comment by jfrazelle** [comment]
this is awesome :) so nice to see everyone working together ;)

i think we just need a rebase then we can get this in and add support for other things in seperate PRs so this can be in for the release ;)

**Comment by jfrazelle** [comment]
wdyt?

**Comment by tianon** [comment]
Oh nice, scraping this information directly from the LSB data is an excellent touch! :+1:

**Comment by fmoliveira** [comment]
Sure @jfrazelle . I'll rebase this tonight (BRT, just a few hours from now) and amend it so we're able to merge. :)

Thanks for your thoughts @tianon !

**Comment by fmoliveira** [comment]
Thanks a million everyone for helping us to test this script. I've just rebased it.

The script is fail-proof for distros which doesn't have the `-u` option. If this still won't work for your distro, please open a new issue.

ping @jfrazelle @benmccann @Djelibeybi @fabioxgn @arner @thaJeztah @sudohippie @tiborvass @tianon

**Comment by Djelibeybi** [comment]
:+1:

**Comment by benmccann** [comment]
:+1:

**Comment by sudohippie** [comment]
:+1:

**Comment by thaJeztah** [comment]
LGTM great work all!

ping @jfrazelle @tianon for review

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
ping @tianon for any last minute words/approval :)

**Comment by thaJeztah** [comment]
:tada: great work!

@calavera think you want this?

**Comment by fmoliveira** [comment]
Thanks everybody for the support in testing and merging this! :tada: :tada:

**Comment by jfrazelle** [comment]
thank you for making something super exstensible!!!

**Comment by fmoliveira** [comment]
You're welcome! ;)

Woot!

**Comment by fmoliveira** [comment]
You're welcome @jfrazelle ! ;)

Woot!

**Comment by fmoliveira** [comment]
You're welcome @jfrazelle ! ;)

It rocks working with all of you!

Woot!

**Comment by gospo** [comment]
Sorry for the delayed response, but it looks like Cumulus Linux doesn't work as the '-u' option isn't supported for lsb_release (same as what was reported with an older version of Mint).  That sucks as I liked this fix a bit more than dragging out the debian line to include more and more forks.

**Comment by calavera** [comment]
cherry picked into #15091

**Comment by fmoliveira** [comment]
@gospo I see. It's a pity the LSB check won't work for Cumulus. I suggest you to open a new issue to discuss with more people about ideas to solve this. There are some stuff I'd ask to try to help but this isn't the ideal place.

**Comment by fmoliveira** [comment]
Closes #14999.
